src/main/java/backtrace/io/BacktraceQueue.java

```diff
@@ -33,32 +40,46 @@ void addWithLock(BacktraceMessage message) {
     void unlock() {
         LOGGER.debug("Releasing semaphore..");
 
-        if(notEmptyQueue.getCount() == 0) {
-            notEmptyQueue.countUp();
+        if (processingLock.getCount() == 1) {
+            processingLock.countDown();
         }
+    }
+
+    boolean shouldHandleMessages() {
+        return processingLock.getCount() == 1;
+    }
 
-        if (lock.getCount() == 0) {
-            return;
+    /**
+     * Lock semaphore because queue is empty
+     */
+    void queueIsEmpty() {
+        LOGGER.debug("Queue is empty, notEmptyQueue counter: " + notEmptyQueueLock.getCount());
+        if (notEmptyQueueLock.getCount() == 0) {
+            LOGGER.debug("Queue is empty - locking thread semaphore");
+            notEmptyQueueLock.countUp();
         }
+    }
 
-        lock.countDown();
+    /**
+     * Unlocking processing because queue is not empty
+     */
+    private void queueNotEmpty() {
+        LOGGER.debug("Queue is NOT empty, notEmptyQueue counter: " + notEmptyQueueLock.getCount());
+        if (notEmptyQueueLock.getCount() == 1) {
+            LOGGER.debug("Queue is not empty - releasing semaphore");
+            notEmptyQueueLock.countDown();
+        }
     }
 
     /**
      * Lock semaphore to inform that at least one of messages are processing
      */
     private void lock() {
-        if(notEmptyQueue.getCount() == 1) {
-            notEmptyQueue.countDown();
-        }
-
-        if (lock.getCount() != 0) {
-            return;
+        if (processingLock.getCount() == 0) {
```

Comment:
shouldn't we use `shouldHandleMessages` function instead?

`if (!shouldHandleMessages()) {`

Reply:
same as above